backend: Use user IDs as recipients.

eeshangarg opened the pull request:
@showell, @timabbott: I had to submit another PR because apparently, GitHub doesn't let me re-order history if the branch has already been pushed to a remote once! So I thought, pushing on a new branch might fix that, but that doesn't cut it either. It might reflect the intended ordering of commits once you fetch the PR locally. :/

@showell: The idea is that after the `extract_recipients` commit, I will add another one on top that finally implements the backend support for using IDs in typing notifications. I had to commit the `extract_recipients` work first because it made testing the backend changes a whole lot easier. I'll push the rest of the changes once they are complete. Getting `mypy` to pass took a lot of work!

After that, the idea is that all commits on top of the last one will be geared towards implementing user ID support for sending PMs and what not!

Thanks! :)

showell commented on zerver/tests/test_messages.py:
```diff
@@ -383,6 +382,29 @@ def assert_invalid_email() -> Any:
+        result_user_ids = [user_profiles[0].id, user_profiles[1].id,
+                           user_profiles[2].id]
+
+        self.assertEqual(result_user_ids, user_ids)
```
I'd set-ify these unless order really matters and is guaranteed.

showell commented on zerver/views/messages.py:
```diff
@@ -295,7 +295,7 @@ def by_pm_with(self, query: Query, operand: str, maybe_negate: ConditionTransfor
                 recipient = recipient_for_emails(emails, False,
                                                  self.user_profile, self.user_profile)
-            except ValidationError:
+            except JsonableError:
```
This is a little unfortunate, since `ValidationError` still would have made sense in this particular context.  Part of me is wondering if we should just fully decouple `recipient_for_emails` from the `Addressee` code.

showell commented on zerver/lib/actions.py:
```diff
@@ -1812,6 +1812,24 @@ def recipient_for_emails(emails: Iterable[str], not_forged_mirror_message: bool,
+
+    return recipient_for_user_profiles(
+        user_profiles=user_profiles,
+        not_forged_mirror_message=not_forged_mirror_message,
```
I don't think we need `not_forged_mirror_message` or `forwarded_user_profile` here, since typing indicators are Zulip-specific and not used in "mirror" scenarios.

showell commented on zerver/lib/actions.py:
```diff
@@ -1875,12 +1877,27 @@ def extract_recipients(s: Union[str, Iterable[str]]) -> List[str]:
+            # if they are not an integer ID
+            recipients.append(recipient)
+
+    is_heterogenous = (not all(isinstance(recipient, int) for recipient in recipients) and
```
I'm still not convinced we need this `is_heterogenous` complexity.  In my mind, we do this:

- split the list
- if empty, return an error
- look at first element, and based on its type, call either `validate_user_ids` or `validate_emails` with lists of strings

Then in each validation function just enforce that elements are either ints or emails, respectively.

showell commented on zerver/lib/addressee.py:
```diff
@@ -105,6 +105,9 @@ def legacy_build(sender: UserProfile,
 
             return Addressee.for_stream(stream_name, topic_name)
         elif message_type_name == 'private':
+            if not message_to:
```
I think this check should be in `for_emails`, so that you can strip and split `message_to` as needed.  You want to guard against `message_to` being something like "    " (which evaluates to `True` in python, I believe, but is still empty from our perspective).

eeshangarg replied:
Okay, I'm not sure if I understand, why should this be in `recipient_for_emails`? This piece of code has nothing to do with that function (I think), right? Also, in Python, I think an empty string evaluates to `False`.

eeshangarg replied:
Also, btw, all of the stripping is already done by `extract_recipients` anyway.

timabbott replied:
I believe this commit is incorrect as written.  Not everything goes through `legacy_build`, and we don't have any explicit validation that the recipients list is non-empty in the `Addressee.for_private()` code path.  I think we just want to add an `assert len(message_to) > 0` sort of thing at the top of `Addressee.for_private` (I think it's OK to 500 in that case, because callers should be doing the appropriate validation), and do an audit of the callers to ensure they can never be empty.  By my read, there's exactly one that can be:

* `send_message` in `bot_lib.py` calls  `internal_send_huddle_message` with a potentially 0-length list.  I think we should fix that one inside `bot_lib.py`; it should be validating its input and returning an error in some appropriate fashion.
* (The other call to `internal_send_huddle_message` in production code is safe, because it starts with a valid Recipient object)

So we should fix that (changing `bot_lib.py` adding the `assert`), and then this commit is mergable.

timabbott replied:
Just bumpinb this.

showell commented:
Looks good so far.  I added a few comments.  You should try to figure out which commit is breaking the build.  The build is reporting that we are losing line coverage on actions.py, which I suspect is due to the `recipients_for_user_ids` changes, but I haven't checked.  That's the one commit where I think we still need a re-think, so it might be worth trying to get a PR with everything up until that.  In particular it would be nice to get a PR merged that does the following:

* break out new function to validate typing indicators
* inline helper into recipient_from_emails
* unsplit get_user_profiles

I can make the JS-side changes to `typing.js` as needed.

eeshangarg commented:
@showell: I just updated this with the following:

* Responded to your critique about not using `mirroring` arguments in `recipient_for_user_ids`.
* Responded to your critique about `extract_recipients`. I have one concern. I don't think we should be complaining about empty recipient lists inside `extract_recipients`. That would mess up how such errors are currently raised throughout the code base, wouldn't it? Isn't it better to complain about empty recipients in `Addressee`, like one of my commit already does? For now, we just return the empty list. Let me know what you think. :)
* I also added another commit that finally adds support for user IDs to `check_send_typing_notification`.

Of course, GitHub messes up the order of the commits. But if you rebase my PR locally, I think the order of the commits would make more sense! :)

Looking forward to your feedback! Thanks! :)

showell commented on zerver/lib/actions.py:
```diff
@@ -1873,12 +1875,39 @@ def extract_recipients(s: Union[str, Iterable[str]]) -> List[str]:
+    # Remove any duplicates.
+    return list(set(recipients))
+
+def validate_user_ids(user_ids: Union[Iterable[str], Iterable[int]]) -> List[int]:
```
I think we should call theses helpers `extract_user_ids` and `extract_emails`.  I don't like `validate_` as a prefix here, since it suggests that we're only validating, whereas we're actually converting data.

eeshangarg commented:
@showell: Just updated this! I'd really appreciate it if you could please give this a full review when you have time! I'd love to see part of this PR getting merged slowly before I stack more commits on top! Thanks! :)

timabbott commented:
`get_user_profile_by_id_in_realm` doesn't include cross-realm users; my guess is that this PR breaks sending messages to `welcome-bot` (for example).

eeshangarg commented:
@timabbott: I just submitted #10463, which implements `get_user_by_id_in_realm_including_cross_realm`. Once we close that PR, I should be able to update this one and @adnrs96 should be able to continue his work on #10360. Thanks! :)

showell commented:
@timabbott I think we can merge everything up to but not include the commit `extract_recipients: Support user IDs' to start chipping away at this, after you look at them.  After that I'm not necessarily saying there are problems, just saying the first few are pretty straightforward.

eeshangarg commented:
@showell, @timabbott: Just updated this with a commit that adds support for user IDs to `handle_deferred_message`. Thanks! :)

eeshangarg commented:
@timabbott: Just rebased and resolved conflicts! Thanks! :)

timabbott commented:
I merged the first 3 commits, since they look great.  One thing I'm curious for your thinking on is why you thought "require them to be all IDs or all emails", was more cumbersome than processing a mix -- can you talk about any of the issues you see with changing that?  I think that would simplify the typing issues (since you'd have an `Iterable[Union[str, int]]`), and I can imagine a use case for it (though not any really good ones).  

For your question above, `create_mirrored_message_users` sorta fundamentally needs to receive emails (the whole point is to create user entries for emails that don't have accounts yet.

eeshangarg commented:
@timabbott: Okay, so initially, I had written `extract_recipients` to be able to support heterogeneous recipient lists. That was some time ago, so I can't exactly remember what my rationale was for asserting that supporting heterogeneous lists is too "cumbersome". So, I just spent a few hours today trying to play around with heterogeneous recipient lists and re-writing a couple of commits. A couple of things I'd like to say:

* Keeping lists homogenous had the added benefit of simplifying things by having logically separate functions such as `recipient_for_user_ids` and `recipient_for_emails` which could operate under the assumption that the lists would either contain only IDs or only emails. Other examples include `Addressee.for_user_ids` and `get_user_profiles_by_ids`. Please note that if we decide to support heterogenous lists, we will have to modify the commits that you have already merged in this PR!
* The typing issues are only slightly improved by supporting heterogeneous lists. The `mypy` annotations for `converter` functions will still have to be modified regardless of whether we support mixed lists or not! I already rewrote the "extract_recipients" commit to support mixed lists, and it doesn't improve our typing situation at all.
* I am really not sure that type checking is less complex with mixed lists overall.

I guess the one takeaway from this explanation should be that supporting mixed lists definitely leads to an implementation that looks somewhat like a "kludge", where we'll just end up adding additional type checking to existing functions. On the other hand, with homogeneous lists, things are a bit cleaner because we can delegate more to functions that can handle either IDs or emails, but not both!

Now, having said that, homogeneous lists complicate things for mirrored users, where they don't have any IDs, which, admittedly, I hadn't thought of! So we have two choces, either do the bulk of the work to support mixed lists (which involves rewriting much of this PR, including the commits that have already been merged), or enforce that recipient lists with mirrored users only use emails. The latter option is definitely less work!

Would love to know your thoughts on how to move forward! Thanks!

@showell: FYI :)

timabbott commented on zerver/lib/actions.py:
```diff
@@ -1876,12 +1878,39 @@ def extract_recipients(s: Union[str, Iterable[str]]) -> List[str]:
-    return list(set(recipient for recipient in recipients if recipient))
+    if isinstance(data[0], str):
+        if data[0].strip().isdigit():
+            recipients = extract_user_ids(data)  # type: Union[List[str], List[int]]
```
Two questions about this.  First, we need a try/except block here, since otherwise `extract_user_ids` can cause us to 500 if the input is mixed.

Second, do we actually need to support string-format user IDs?  I don't see a reason to allow that, in which case we can just delete this block.

timabbott replied:
(There's a good use case for being fast-and-loose with comma-separated emails, for convenience of typing, but with user IDs, you're asking the client to marshall the data into a format for you, so you might as well enforce ti being the right format)

timabbott commented on zerver/tests/test_typing.py:
```diff
@@ -78,6 +78,43 @@ def test_single_recipient(self) -> None:
+        is successful
+        """
+        sender = self.example_user('hamlet')
+        recipient = self.example_user('othello')
```
call this `recipient_user`; we generally reserve `recipient` for Recipient objects.

eeshangarg replied:
Done :)

timabbott commented on zerver/tests/test_messages.py:
```diff
@@ -1924,6 +1924,26 @@ def do_schedule_message(self, msg_type: str, to: str, msg: str,
+            forwarder_user_profile=self.example_user('hamlet'),
+            realm=None
+        )
+        self.assert_json_success(result)
```
I don't understand this commit; can you explain why you're adding this brittle test?

eeshangarg replied:
Oops that's my bad, this commit message is supposed to say `mypy: Accept user IDs in handle_deferred_message.` (which is what the commit is called locally on my branch so I just forgot to push probably) and is supposed to prevent mypy errors because callers may pass in integer user IDs to these functions now.

The brittle test shouldn't be there either. Initially, I had thought that I would not add support for user IDs to these methods and only focus on `send_message_backend` first. So I went down the rabbit hole of doing a manual type cast inside `handle_deferred_message`, and that's when I added the test. Anyway, updating now! :)

timabbott commented on zerver/views/messages.py:
```diff
@@ -1252,6 +1253,9 @@ def send_message_backend(request: HttpRequest, user_profile: UserProfile,
             return json_error(_("User not authorized for this query"))
+        # For now, mirroring only works with recipient emails, not for
+        # recipient user IDs.
+        message_to = cast(Sequence[str], message_to)
```
This should be enforced by an actual check, not just a typing system cast.

timabbott replied:
e.g. `if not all(isinstance(to_item, str) for to_item in message_to):`

eeshangarg replied:
Okay, so I added the actual check that enforces this, but we still need the manual cast after that check so that `mypy` doesn't complain about `create_mirrored_send_message` not being able to accept a `Sequence[int]` type parameter.

timabbott commented on zerver/tests/test_messages.py:
```diff
@@ -1395,6 +1395,22 @@ def test_personal_message(self) -> None:
+                "to": ujson.dumps([self.example_user("othello").id])
+            }
+        )
+        self.assert_json_success(result)
```
Let's include a `get_last_message` check that it actually sent a PM to Othello.

eeshangarg replied:
Done! :)

timabbott replied:
This should also have a test for group PMs; I'll just add that while merging.

timabbott commented:
I've decided I'm fine with merging this with the homogenous approach.  Posted a bunch of comments on issues that can be fixed, and then this should be mergable.

eeshangarg commented:
@timabbott: Just updated this and addressed all of your comments! Looking forward to your feedback! Thanks! :)

timabbott commented on zerver/tests/test_typing.py:
```diff
@@ -136,6 +203,40 @@ def test_send_notification_to_self_event(self) -> None:
+        self.assertEqual(event_recipient_emails, expected_recipient_emails)
+        self.assertEqual(event['sender']['email'], email)
+        self.assertEqual(event['type'], 'typing')
+        self.assertEqual(event['op'], 'start')
```
I'm going to merge this, but I think given how repetitive these tests are, it's probably worth extracting a helper function for the common code as a follow-up.

timabbott commented:
Merged, thanks @eeshangarg!  Congrats on finishing a rather large project.  There's a few follow-ups we should probably invest in:
* Support sending to streams by ID, which should be a much simpler version of very similar work.
* Updating our API documentation to document support for this in the `send_message` API call (and, if necessary, updating the Python+JS bindings to support passing IDs)
* The testing deduplication cleanup for typing notifications I mentioned above.
